src/ice_nucleation.F90

```diff
@@ -0,0 +1,563 @@
+! Copyright (C) 2024 University of Illinois at Urbana-Champaign
+! Licensed under the GNU General Public License version 2 or (at your
+! option) any later version. See the file COPYING for details.
+
+!> \file
+! The pmc_ice_nucleation module.
+
+module pmc_ice_nucleation
+  use pmc_aero_state
+  use pmc_env_state
+  use pmc_aero_data
+  use pmc_util
+  use pmc_aero_particle
+  use pmc_constants
+  use pmc_rand
+
+  implicit none
+
+  !> Type code for an undefined or invalid immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_INVALID = 0
+  !> Type code for constant ice nucleation rate (J_het) immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_CONST = 1
+  !> Type code for the singular (INAS) immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_SINGULAR = 2
+  !> Type code for the ABIFM immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_ABIFM = 3
+
+  !> Whether to use the binned-tau leaping algorithm for time-dependent scheme.
+  logical :: do_speedup = .true.
+
+  interface ABIFM
+    module procedure ABIFM_particle
+    module procedure ABIFM_max
+  end interface
+
+contains
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Main subroutine for immersion freezing simulation.
+  subroutine immersion_freezing(aero_state, aero_data, env_state_initial, &
```

Comment:
This subroutine should be named something with the file name in it like we do for other processes (such as `ice_nucleation_immersion_freezing`.